### PR TITLE
[#5919] Platform: change default storage option for AWS provider to 1x250GB

### PR DESCRIPTION
**Pull request by jitendra-12113**
**Heading:**
[#5919] Platform: change default storage option for AWS provider to 1x250GB
**Description:**
The empty volume list will be stored in the db in that case we will be using default value(volume_count, volume_size configured in refernce.conf. 
**Testing:**
When we select aws provider while creating universe, volume count was showing more than one for some instances types starting with ""c5./c4.". Now with this fix it will show only one.
Will have to reset instanceTypeDetailsJson to empty for older YW
instances.
Added UT's as well. sbt test run results:

[info] Test com.yugabyte.yw.commissioner.TaskGarbageCollectorTest.testPurge_noneStale started
[info] Test com.yugabyte.yw.commissioner.TaskGarbageCollectorTest.testIgnorePromError started
[info] Test com.yugabyte.yw.commissioner.TaskGarbageCollectorTest.testPurge_invalidData started
[info] Test run finished: 0 failed, 0 ignored, 7 total, 0.589s
[info] Passed: Total 1092, Failed 0, Errors 0, Passed 1088, Skipped 4
[success] Total time: 575 s, completed 18 Jan, 2021 12:33:44 PM
hasher@LAP-LIN-908:~/yugabyte-db/managed$

### CODE REVIEW

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:53]: `Config` not `ConfigFactory`
Like I said use TaskGarbageCollector as example.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:36]: why `static`?

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:197]: Inject config. Then no need to `load`.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:34]: YB_VOLUME_INFO_VOLUME_COUNT_KEY

**Review by sb-yb — CHANGES_REQUESTED** [review]

**Comment by sb-yb** [comment]
Also bug description should say what we are doing here.
Basically the fact that empty volume list is stored in db when we want to use config default.
Also Fix the "Heading" 
It should be following format 
`[#1234] Platform: One line description of change`

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:194]: extract method `populateDefaultIfEmpty()`

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:53]: With Config,  always getting NLP(Null pointers), hence used ConfigFactory instead of.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:202]: done

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: changed config var to non-static, others kept as it is, since those are being used in static block.

Comment (sb-yb) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: yep. my comment was only about config.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:53]: Thats odd. In tests or when running server? May be some ordering issue during binding.
Inject 
`private final Provider<Config> config`
@Inject
InstanceType(Provider<Config> config) {
...
}`

then where you use it do 

`config.get()`

This delays the resolution to runtime.

Let me know if that fixes the problem. If not will patch and see whats wrong.

The config factory is not an option here. That code is buggy. 
Another way is to get to typesafe config through play but lets try this first.

File: managed/src/main/resources/reference.conf
Comment (sb-yb) [inline managed/src/main/resources/reference.conf:29]: `# default volume count for aws instance types with EBS Only  storage info`

File: managed/src/main/resources/reference.conf
Comment (sb-yb) [inline managed/src/main/resources/reference.conf:30]: `default_`

File: managed/src/main/resources/reference.conf
Comment (sb-yb) [inline managed/src/main/resources/reference.conf:33]: `default_`

File: managed/src/main/resources/reference.conf
Comment (sb-yb) [inline managed/src/main/resources/reference.conf:28]: `aws {
...
}`

**Review by sb-yb — CHANGES_REQUESTED** [review]
Few more changes.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: Its happening while running server. 
Getting below error:
Caused by: com.google.inject.CreationException: Unable to create injector, see the following errors:
1) Error injecting constructor, java.lang.NullPointerException
  at AppInit.<init>(AppInit.java:47)
  at Module.configure(Module.java:48) (via modules: com.google.inject.util.Modules$OverrideModule -> Module)
  while locating AppInit
  
  Tried above things as well but that didn't work out. 
  Updated a patch, using config only. Please suggest! Thanks!

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: done

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: Done.

File: managed/src/main/resources/reference.conf
Comment (jitendra-12113) [reply managed/src/main/resources/reference.conf]: Done

File: managed/src/main/resources/reference.conf
Comment (jitendra-12113) [reply managed/src/main/resources/reference.conf]: Done

File: managed/src/main/resources/reference.conf
Comment (jitendra-12113) [reply managed/src/main/resources/reference.conf]: Done

File: managed/src/main/resources/reference.conf
Comment (jitendra-12113) [reply managed/src/main/resources/reference.conf]: Thanks! Done.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: So I saw a `provider` for `Config` in `Module`. But looks like that is a different Config type not the `typesafe.Config`. 
So it is not going to work this way. Provider trick also does not delay the resolution because we need it during AppInit.

So way to do this gracefully is 
 - Inject `Application` in `InstanceType`
  - Then you can get to typesafe config using `app.config()`  (NOT `app.configuration()` which is deprecated)
   `this.config = app.config()`
   
   Let me know if that helps.

Comment (sb-yb) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: Just noticed that InstanceType is a Entity bean. So it wont use guice injection. It is using the default ctor (that you added) followed by the field mapping by the ORM  and thus config remains null.
So you can just pass the config value into findByProvider as parameter. No config field.

Sorry for confusion. Right way to do this is not use so much static factory methods. Usually there is a singleton helper class like InstanceTypeHelper. That will have non-static findByProvider method. then you can actually do proper guice injection in such class.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: If this is the only config  value then pass a parameter in to this method `awsDefaultVolumeCount` 
Do config lookup in caller.
I will do a refactor later which should fix this DI fail as described above - using singleton helper classes.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: Done. Added this for now, later we will refactor the code once we have Helper method in place as you suggested above. Thanks!

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: Done!

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:203]: Do not make separate method. Just update same method.

**Review by sb-yb — CHANGES_REQUESTED** [review]
Add unit tests

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:203]: @sb-yb  I thought of doing this way but we will have to update it every place like AppInit.java and other places wherever  this method is being used. 
We know this is temp fix, anyway we are going to refactor the code. That's why created overloaded method to avoid changing everywhere.

File: managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java:35]: Here we will have to use Inject, Without inject we are getting below error.

1) No implementation for com.yugabyte.yw.controllers.InstanceTypeController (with no qualifier annotation) was bound, and could not find an injectable constructor. Injectable classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.

File: managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java]: done

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: Updated the same method instead of having overloaded method.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/models/InstanceType.java]: Updated! Thank you!

**Review by sb-yb — CHANGES_REQUESTED** [review]
@jitendra-12113 anything holding up this change? Do you need help with any part?

**Comment by jitendra-12113** [comment]
> @jitendra-12113 anything holding up this change? Do you need help with any part?

@sb-yb  Have updated the patch. I think we are good to go ahead and merge this change. Need your approval though. Thanks!

File: managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java:41]: Do not mix ctor injections and field injections in same class.
Either fix all existing fields to be parameters here or make the new field use field injection.

FWIW, using ctor injection with private fina fields is the best practice. But i ll leave it to you which option you want for now.

File: managed/src/main/java/com/yugabyte/yw/models/InstanceType.java
Comment (sb-yb) [inline managed/src/main/java/com/yugabyte/yw/models/InstanceType.java:190]: Instead consider moving the whole code (line 190 to 196) in populateDefaultIfEmpty() 
And name the method populateDefaultsIfEmpty (plural)

File: managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java
Comment (sb-yb) [inline managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java:34]: where is a test where you have to setup this mock?
`when(mockConfig.get(..._COUNT)).thenReturn(1);`
`when(mockConfig.get(..._GB)).thenReturn(250);`

**Review by sb-yb — CHANGES_REQUESTED** [review]
Final lap:
small code change and missing unit test

File: managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java
Comment (jitendra-12113) [reply managed/src/main/java/com/yugabyte/yw/controllers/InstanceTypeController.java:38]: Done. Thanks!

File: managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java
Comment (jitendra-12113) [reply managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java]: Ah! Didn't realise this is for entity class. We can't mock config in this file. Most of the values are being read from in memory.
Instead used ConfigFactory.load() just to test our code. I think, this should be fine if not please suggest! 
Also due to static method used in Entity class, getting difficulties in writing Unit test cases, need to refactor the code(Entity class). But for now I think this should be fine. 

Thanks!

File: managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java
Comment (sb-yb) [inline managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java:96]: This is incorrect. Not sure what is your confusion.
Here is what you have to do in great detail:
1. Revert to using mock.Config 
2. See what is in defaultDetails  - line 44 populates it. But we do not want to test that code path. To excercise our code path the defaultDetails has to be empty. Thats when the populateEmpty will do something and try to lookup config values.
3. Like line 88 upsert another instancetype  such that its details have volumeList empty.
4. Now the test will fail complaining that you have not mocked config.getInt or whatever you use. 
5. Add mock setup. For example volume count config return 3 and size returns 100GB.
6.  assert that the volume list for this instance as returned by findPyProvider is size 3 and each one is 100GB.

Hope it helps.

**Review by sb-yb — CHANGES_REQUESTED** [review]
Given detailed instructions inline. LMK if it helps.

File: managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java
Comment (jitendra-12113) [reply managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java:101]: Had to refactor the code lil bit. Forgot to apply @RunWith(MockitoJUnitRunner.class) on top to use mock. Now its working fine. I really learnt a lot while working on this change.  Please let me know if you still thinks, some changes are required. Thanks for all your suggestions:)

Comment (sb-yb) [reply managed/src/test/java/com/yugabyte/yw/models/InstanceTypeTest.java:101]: 💯  for right attitude!

**Review by sb-yb — APPROVED** [review]
This looks great now! 
Important tips:
Make sure to run the unittests before submit. If any fail take it seriously and rerun failed tests. 
If you run sbt shell then you can do `testQuick` and most of the times that only re-runs failed tests.  Or you can use testOnly for specific tests. There are few flaky tests so you can ignore if they pass in subsequent run. If not then take it seriously and post on slack channel for any debugging help. Sometime there are broken tests at HEAD. Not ideal situation, I know.  But given that it is easier to not spend too much time debugging and just ask around (I learned painfully :( ..)

**Pull request merged**
